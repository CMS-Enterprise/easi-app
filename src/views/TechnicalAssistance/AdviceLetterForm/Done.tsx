import React from 'react';
import { useTranslation } from 'react-i18next';
import { useHistory, useLocation, useParams } from 'react-router-dom';
import { ButtonGroup, Grid, GridContainer } from '@trussworks/react-uswds';
import classNames from 'classnames';

import UswdsReactLink from 'components/LinkWrapper';
import PageHeading from 'components/PageHeading';
import cmsDivisionsAndOffices from 'constants/enums/cmsDivisionsAndOffices';
import useTRBAttendees from 'hooks/useTRBAttendees';

import Breadcrumbs from '../Breadcrumbs';

/**
 * Last form step for confirmation
 */
const Done = () => {
  const { t } = useTranslation('technicalAssistance');

  const { id } = useParams<{ id: string }>();
  const { state } = useLocation<{ success?: boolean }>();
  const history = useHistory();

  const {
    data: { requester }
  } = useTRBAttendees(id);

  const success = state?.success;

  // Redirect to first step if success state is undefined
  if (success === undefined) {
    history.replace(`/trb/${id}/advice/summary`);
  }

  return (
    <>
      <div
        className={classNames(
          {
            'bg-success-lighter': success,
            'bg-error-lighter': !success
          },
          'padding-bottom-6'
        )}
      >
        <GridContainer>
          <Breadcrumbs
            items={[
              { text: t('Home'), url: '/trb' },
              {
                text: t(`Request ${id}`),
                url: `/trb/${id}/advice`
              },
              { text: t('adviceLetterForm.heading') }
            ]}
          />
          <PageHeading className="margin-bottom-0">
            {t(`done.${success ? 'success' : 'error'}.heading`)}
          </PageHeading>
          <p className="font-body-lg line-height-body-5 text-light margin-top-05">
            {t(`adviceLetterForm.done.${success ? 'success' : 'error'}`)}
          </p>
          <ButtonGroup>
            {!success && (
              <UswdsReactLink
                to={`/trb/${id}/advice/review`}
                className="usa-button"
              >
                {t('adviceLetterForm.done.backToAdviceLetter')}
              </UswdsReactLink>
            )}
            <UswdsReactLink
              to={`/trb/${id}/request`}
              className={classNames('usa-button', {
                'usa-button--outline': !success,
                'margin-left-neg-05': success
              })}
            >
              {t('adviceLetterForm.done.returnToRequestHome')}
            </UswdsReactLink>
          </ButtonGroup>
        </GridContainer>
      </div>

      <GridContainer className="margin-top-4">
        <dl className="grid-row easi-dl">
          <Grid col={6}>
            <dt>Request Name</dt>
            <dd>{id}</dd>
          </Grid>
          <Grid col={6}>
            <dt>Requester</dt>
            <dd>{`${requester.userInfo?.commonName}, ${
              cmsDivisionsAndOffices.find(
                component => component.name === requester.component
              )?.acronym
            }`}</dd>
          </Grid>
        </dl>
      </GridContainer>
    </>
  );
};

export default Done;
