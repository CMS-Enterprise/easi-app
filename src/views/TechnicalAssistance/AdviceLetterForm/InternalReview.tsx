import React, { useEffect } from 'react';
import { Trans, useTranslation } from 'react-i18next';
import { useHistory } from 'react-router-dom';
import { ApolloError, useMutation } from '@apollo/client';
import { Button } from '@trussworks/react-uswds';

import Alert from 'components/shared/Alert';
import Divider from 'components/shared/Divider';
import {
  DeleteTrbRecommendationQuery,
  GetTrbAdviceLetterQuery,
  RequestReviewForTRBAdviceLetterQuery
} from 'queries/TrbAdviceLetterQueries';
import {
  DeleteTRBRecommendation,
  DeleteTRBRecommendationVariables
} from 'queries/types/DeleteTRBRecommendation';
import {
  RequestReviewForTRBAdviceLetter,
  RequestReviewForTRBAdviceLetterVariables
} from 'queries/types/RequestReviewForTRBAdviceLetter';
import { TRBAdviceLetterStatus } from 'types/graphql-global-types';
import { StepComponentProps } from 'types/technicalAssistance';

import ReviewAdviceLetter from '../AdminHome/components/ReviewAdviceLetter';
import Pager from '../RequestForm/Pager';

const InternalReview = ({
  trbRequestId,
  adviceLetter,
  adviceLetterStatus,
  setFormAlert,
  setIsStepSubmitting,
  stepsCompleted,
  setStepsCompleted
}: StepComponentProps) => {
  const { t } = useTranslation('technicalAssistance');
  const history = useHistory();

  const [requestReview, { loading: isSubmitting }] = useMutation<
    RequestReviewForTRBAdviceLetter,
    RequestReviewForTRBAdviceLetterVariables
  >(RequestReviewForTRBAdviceLetterQuery, {
    variables: {
      id: adviceLetter.id
    }
  });

  const [remove] = useMutation<
    DeleteTRBRecommendation,
    DeleteTRBRecommendationVariables
  >(DeleteTrbRecommendationQuery, {
    refetchQueries: [
      {
        query: GetTrbAdviceLetterQuery,
        variables: {
          id: trbRequestId
        }
      }
    ]
  });

  useEffect(() => {
    setIsStepSubmitting(isSubmitting);
  }, [setIsStepSubmitting, isSubmitting]);

  return (
    <div id="trbAdviceInternalReview">
      {/* Internal Review */}
      <ReviewAdviceLetter
        trbRequestId={trbRequestId}
        adviceLetter={adviceLetter}
        className="margin-top-5 margin-bottom-4"
        recommendationActions={{
          edit: {
            onClick: recommendation =>
              history.push(`/trb/${trbRequestId}/advice/recommendations/form`, {
                recommendation: {
                  ...recommendation,
                  links: recommendation.links.map(link => ({ link }))
                }
              })
          },
          remove: {
            onClick: recommendation =>
              remove({ variables: { id: recommendation.id } }).catch(() =>
                setFormAlert({
                  type: 'error',
                  message: t('adviceLetterForm.error', {
                    action: 'removing',
                    type: 'recommendation'
                  })
                })
              )
          }
        }}
        showSectionEditLinks
      />

      <Divider />

      {/* Internal review needed alert */}
      <Alert
        type="warning"
        heading={t('adviceLetterForm.internalReviewNeeded.heading')}
        className="margin-top-4 margin-bottom-5"
      >
        {t('adviceLetterForm.internalReviewNeeded.text')}
      </Alert>

      {/* Form pager buttons */}
      <Pager
        back={{
          outline: true,
          onClick: () => history.push(`/trb/${trbRequestId}/advice/next-steps`)
        }}
        next={{
          text: t(
            `adviceLetterForm.${
              adviceLetterStatus === TRBAdviceLetterStatus.IN_PROGRESS
                ? 'requestInternalReview'
                : 'requestAnotherInternalReview'
            }`
          ),
          disabled: isSubmitting,
          onClick: () => {
            requestReview()
              .then(() => history.push(`/trb/${trbRequestId}/advice`))
              .catch(error => {
                if (error instanceof ApolloError) {
                  setFormAlert({
                    type: 'error',
                    message: t('adviceLetterForm.error', {
                      action: 'submitting',
                      type: 'advice letter for internal review'
                    })
                  });
                }
              });
          }
        }}
        buttons={[
          <Trans i18nKey="technicalAssistance:adviceLetterForm.progressToNextStep">
            one
            <Button
              type="button"
              unstyled
              onClick={() => {
                if (
                  setStepsCompleted &&
                  stepsCompleted &&
                  !stepsCompleted?.includes('internal-review')
                ) {
                  setStepsCompleted([...stepsCompleted, 'internal-review']);
                }
                history.push(`/trb/${trbRequestId}/advice/review`);
              }}
            >
              button
            </Button>
          </Trans>
        ]}
        taskListUrl={`/trb/${trbRequestId}/advice`}
        saveExitText={t('adviceLetterForm.returnToRequest')}
        submitDisabled
        border={false}
      />
    </div>
  );
};

export default InternalReview;
