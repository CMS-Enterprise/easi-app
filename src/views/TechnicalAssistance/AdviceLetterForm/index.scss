.usa-form .usa-form-group .usa-fieldset.text-normal legend.usa-legend {
  font-weight: normal;
}
