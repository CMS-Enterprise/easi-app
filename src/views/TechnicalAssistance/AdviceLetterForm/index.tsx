import React, { useCallback, useEffect, useState } from 'react';
import { useTranslation } from 'react-i18next';
import { useHistory, useLocation, useParams } from 'react-router-dom';
import { useQuery } from '@apollo/client';
import {
  Button,
  Grid,
  GridContainer,
  IconArrowBack
} from '@trussworks/react-uswds';
import { isEqual } from 'lodash';

import PageLoading from 'components/PageLoading';
import { Alert } from 'components/shared/Alert';
import StepHeader from 'components/StepHeader';
import { GetTrbAdviceLetterQuery } from 'queries/TrbAdviceLetterQueries';
import {
  GetTrbAdviceLetter,
  GetTrbAdviceLetterVariables
} from 'queries/types/GetTrbAdviceLetter';
import { TRBAdviceLetterStatus } from 'types/graphql-global-types';
import { FormAlertObject } from 'types/technicalAssistance';
import {
  meetingSummarySchema,
  nextStepsSchema
} from 'validations/trbRequestSchema';
import NotFound from 'views/NotFound';

import Breadcrumbs from '../Breadcrumbs';
import { StepSubmit } from '../RequestForm';

import Done from './Done';
import InternalReview from './InternalReview';
import NextSteps from './NextSteps';
import Recommendations from './Recommendations';
import Review from './Review';
import Summary from './Summary';

import './index.scss';

type StepsText = { name: string; longName?: string; description?: string }[];

export type FormStepKey =
  | 'summary'
  | 'recommendations'
  | 'next-steps'
  | 'internal-review'
  | 'review';

const adviceFormSteps = [
  {
    slug: 'summary',
    component: Summary
  },
  {
    slug: 'recommendations',
    component: Recommendations
  },
  {
    slug: 'next-steps',
    component: NextSteps
  },
  {
    slug: 'internal-review',
    component: InternalReview
  },
  {
    slug: 'review',
    component: Review
  }
] as const;

type AdviceFormStep = typeof adviceFormSteps[number];

/**
 * TRB request admin advice letter form
 */
const AdviceLetterForm = () => {
  const { t } = useTranslation('technicalAssistance');
  const history = useHistory();
  const location = useLocation<{
    error?: boolean;
    state: { fromAdmin?: boolean };
    fromAdmin?: boolean;
  }>();

  const fromAdmin = location.state?.fromAdmin;

  // Get url params
  const { id, formStep, subpage } = useParams<{
    id: string;
    formStep: string;
    subpage: string;
  }>();

  // TRB request query
  const { data, loading } = useQuery<
    GetTrbAdviceLetter,
    GetTrbAdviceLetterVariables
  >(GetTrbAdviceLetterQuery, {
    variables: { id }
  });

  /** Current trb request */
  const trbRequest = data?.trbRequest;
  const { adviceLetter } = trbRequest || {};

  // References to the submit handler and submitting state of the current form step
  const [stepSubmit, setStepSubmit] = useState<StepSubmit | null>(null);
  const [isStepSubmitting, setIsStepSubmitting] = useState<boolean>(false);

  const [stepsCompleted, setStepsCompleted] = useState<FormStepKey[]>();

  // Form level alerts from step components
  const [formAlert, setFormAlert] = useState<FormAlertObject | null>(null);

  /** Form steps translated text object */
  const steps = t<StepsText>('adviceLetterForm.steps', { returnObjects: true });

  /** Index of current form step - will return -1 if invalid URL */
  const currentStepIndex: number = adviceFormSteps.findIndex(
    ({ slug }) => slug === formStep
  );

  /** Current form step object */
  const currentFormStep: AdviceFormStep = adviceFormSteps[currentStepIndex];

  // When navigating to a different step, checks if all previous form steps are valid
  const checkValidSteps = useCallback(
    (index: number): boolean => {
      return (
        adviceFormSteps.filter(
          (step, i) => stepsCompleted?.includes(step.slug) && i < index
        ).length === index
      );
    },
    [stepsCompleted]
  );

  /** Redirect if previous steps are not completed */
  const redirectStep = useCallback(() => {
    if (stepsCompleted && !checkValidSteps(currentStepIndex - 1)) {
      /** Returns latest available step index */
      const stepRedirectIndex = !stepsCompleted.includes('summary')
        ? 0
        : // If summary is completed, return index of last completed step plus 1
          adviceFormSteps.findIndex(
            step => step.slug === stepsCompleted?.slice(-1)[0]
          ) + 1;

      if (!fromAdmin && currentStepIndex === 0) return;
      if (!adviceFormSteps[stepRedirectIndex]?.slug) return;
      // Redirect to latest available step
      history.replace(
        `/trb/${id}/advice/${adviceFormSteps[stepRedirectIndex].slug}`
      );
    }
  }, [
    stepsCompleted,
    currentStepIndex,
    history,
    id,
    checkValidSteps,
    fromAdmin
  ]);

  useEffect(() => {
    if (!adviceLetter) return;
    (async () => {
      let completed: FormStepKey[] = stepsCompleted ? [...stepsCompleted] : [];
      const stepValidators = [];

      // Check the Meeting Summary step
      if (!completed.includes('summary')) {
        const { meetingSummary } = adviceLetter || {};
        stepValidators.push(
          meetingSummarySchema
            .isValid(
              { meetingSummary },
              {
                strict: true
              }
            )
            .then(valid => {
              if (valid) completed = ['summary'];
            })
        );
      }

      // Check the Next Steps step
      if (!completed.includes('next-steps')) {
        const {
          nextSteps,
          isFollowupRecommended,
          followupPoint
        } = adviceLetter;

        stepValidators.push(
          nextStepsSchema
            .isValid(
              { nextSteps, isFollowupRecommended, followupPoint },
              {
                strict: true
              }
            )
            .then(valid => {
              // Internal review should be marked completed with next steps
              if (valid) {
                completed = ['summary', 'recommendations', 'next-steps'];
              }
            })
        );
      }

      if (
        trbRequest?.taskStatuses.adviceLetterStatus ===
          TRBAdviceLetterStatus.READY_FOR_REVIEW &&
        !stepsCompleted?.includes('review')
      ) {
        completed = [
          'summary',
          'recommendations',
          'next-steps',
          'internal-review'
        ];
      }

      Promise.allSettled(stepValidators).then(() => {
        if (!isEqual(completed, stepsCompleted)) {
          setStepsCompleted(completed);
        } else {
          /** Once stepsCompleted is finished evaluating, redirect to last valid step if necessary */
          redirectStep();
        }
      });
    })();
  }, [
    stepsCompleted,
    adviceLetter,
    trbRequest?.taskStatuses?.adviceLetterStatus,
    redirectStep,
    formStep
  ]);

  useEffect(() => {
    if (formAlert) {
      const err = document.querySelector('.trb-form-error');
      err?.scrollIntoView();
    }
  }, [formAlert]);

  useEffect(() => {
    if (!adviceLetter && !loading) {
      const type = location?.state?.error ? 'error' : 'info';
      setFormAlert({
        type,
        message: t(`adviceLetter.alerts.${type}`)
      });
    }
  }, [adviceLetter, loading, location?.state?.error, t]);

  // Page loading
  if (loading) return <PageLoading />;

  // If invalid trb request, show not found
  if (!trbRequest) {
    return <NotFound />;
  }

  const {
    taskStatuses: { adviceLetterStatus }
  } = trbRequest;

  if (formStep === 'done') {
    return <Done />;
  }

  return (
    <>
      {/** Form page header */}
      {!subpage && (
        <StepHeader
          heading={t('adviceLetterForm.heading')}
          text={t('adviceLetterForm.description')}
          subText={t('adviceLetterForm.text')}
          step={currentStepIndex + 1}
          steps={steps.map((step, index) => ({
            key: step.name,
            disabled:
              isStepSubmitting ||
              currentStepIndex === index ||
              !checkValidSteps(index),
            label: (
              <>
                <span className="name">{step.name}</span>
                <span className="long">{step.longName ?? step.name}</span>
              </>
            ),
            description: step.description,
            completed: index < currentStepIndex,
            onClick: async () => {
              const url = `/trb/${id}/advice/${adviceFormSteps[index].slug}`;

              if (stepSubmit) {
                stepSubmit?.(() => history.push(url));
              } else {
                history.push(url);
              }
            }
          }))}
          breadcrumbBar={
            <Breadcrumbs
              items={[
                { text: t('adminHome.home'), url: '/trb' },
                {
                  text: t('adminHome.breadcrumb'),
                  url: `/trb/${id}/advice`
                },
                { text: t('adviceLetterForm.heading') }
              ]}
            />
          }
          errorAlert={
            formAlert && (
              <Alert
                type={formAlert.type}
                className={`trb-form-${formAlert.type} margin-top-3`}
                slim
                closeAlert={setFormAlert}
              >
                {formAlert.message}
              </Alert>
            )
          }
          hideSteps={!adviceLetter}
        >
          {
            /* Save and return to request button */
            !!adviceLetter && (
              <Button
                type="button"
                unstyled
                disabled={isStepSubmitting}
                onClick={() => {
                  const url = `/trb/${id}/advice`;
                  if (stepSubmit) {
                    stepSubmit?.(() => history.push(url), false);
                  } else {
                    history.push(url);
                  }
                }}
              >
                <IconArrowBack className="margin-right-05 margin-bottom-2px text-tbottom" />
                {t('adviceLetterForm.returnToRequest')}
              </Button>
            )
          }
        </StepHeader>
      )}

      {/* Current form step component */}
      <GridContainer>
        <Grid>
          {adviceLetter && (
            <currentFormStep.component
              trbRequestId={id}
              adviceLetter={adviceLetter}
              adviceLetterStatus={adviceLetterStatus}
              setFormAlert={setFormAlert}
              setStepSubmit={setStepSubmit}
              setIsStepSubmitting={setIsStepSubmitting}
              stepsCompleted={stepsCompleted}
              setStepsCompleted={setStepsCompleted}
            />
          )}
        </Grid>
      </GridContainer>
    </>
  );
};

export default AdviceLetterForm;
