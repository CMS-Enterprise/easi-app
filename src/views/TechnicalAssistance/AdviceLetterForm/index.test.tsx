import React from 'react';
import { Provider } from 'react-redux';
import { MemoryRouter, Route } from 'react-router-dom';
import { MockedProvider, MockedResponse } from '@apollo/react-testing';
import {
  render,
  screen,
  waitForElementToBeRemoved
} from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import {
  adviceLetter,
  getTrbAdviceLetterQuery,
  taskStatuses
} from 'data/mock/trbRequest';
import { MessageProvider } from 'hooks/useMessage';
import { CreateTrbRecommendationQuery } from 'queries/TrbAdviceLetterQueries';
import {
  CreateTRBRecommendation,
  CreateTRBRecommendationVariables
} from 'queries/types/CreateTRBRecommendation';
import {
  GetTrbAdviceLetter,
  GetTrbAdviceLetterVariables
} from 'queries/types/GetTrbAdviceLetter';
import { MockedQuery } from 'types/util';
import easiMockStore from 'utils/testing/easiMockStore';
import { mockTrbRequestId } from 'utils/testing/MockTrbAttendees';

import AdviceLetterForm from '.';

const mockRecommendation = {
  trbRequestId: mockTrbRequestId,
  title: 'Recommendation 3',
  recommendation: 'Recommendation description text',
  links: ['google.com', 'easi.cms.gov']
};

const createTrbRecommendationQuery: MockedQuery<
  CreateTRBRecommendation,
  CreateTRBRecommendationVariables
> = {
  request: {
    query: CreateTrbRecommendationQuery,
    variables: {
      input: mockRecommendation
    }
  },
  result: {
    data: {
      createTRBAdviceLetterRecommendation: {
        __typename: 'TRBAdviceLetterRecommendation',
        id: '670fdf6d-761b-415f-a108-2ebc814288c3',
        ...mockRecommendation
      }
    }
  }
};

const defaultStore = easiMockStore({
  euaUserId: 'SF13',
  groups: ['EASI_TRB_ADMIN_D']
});

const getAdviceLetterCannotStart: MockedQuery<
  GetTrbAdviceLetter,
  GetTrbAdviceLetterVariables
> = {
  ...getTrbAdviceLetterQuery,
  result: {
    data: {
      trbRequest: {
        ...getTrbAdviceLetterQuery.result.data?.trbRequest!,
        adviceLetter: null,
        taskStatuses
      }
    }
  }
};

const renderForm = (
  step: string,
  mocks?: MockedResponse[],
  error?: boolean
) => {
  window.HTMLElement.prototype.scrollIntoView = jest.fn;
  return render(
    <MemoryRouter
      initialEntries={[
        {
          pathname: `/trb/${mockTrbRequestId}/advice/${step}`,
          state: { error }
        }
      ]}
    >
      <MessageProvider>
        <Provider store={defaultStore}>
          <MockedProvider
            mocks={
              mocks || [getTrbAdviceLetterQuery, createTrbRecommendationQuery]
            }
            addTypename={false}
          >
            <Route path="/trb/:id/advice/:formStep/:subpage?">
              <AdviceLetterForm />
            </Route>
          </MockedProvider>
        </Provider>
      </MessageProvider>
    </MemoryRouter>
  );
};

const waitForPageLoad = async () =>
  waitForElementToBeRemoved(() => screen.getByTestId('page-loading'));

describe('TRB Advice Letter Form', () => {
  it('matches the snapshot', async () => {
    const { asFragment } = renderForm('summary');

    await waitForPageLoad();

    expect(asFragment()).toMatchSnapshot();
  });

  it('Renders no advice letter alert', async () => {
    const { findByText, findByTestId } = renderForm('summary', [
      getAdviceLetterCannotStart
    ]);

    expect(await findByTestId('alert')).toBeInTheDocument();

    expect(
      await findByText(
        'There is no advice letter for this request yet. Once the consult date has passed, you may create an advice letter for this request.'
      )
    ).toBeInTheDocument();
  });

  it('Renders error creating advice letter alert', async () => {
    const { findByText, findByTestId } = renderForm(
      'summary',
      [getAdviceLetterCannotStart],
      true
    );

    expect(await findByTestId('alert')).toBeInTheDocument();

    expect(
      await findByText(
        'There was an error creating this advice letter. Please try again. If the error persists, please try again at a later date.'
      )
    ).toBeInTheDocument();
  });

  it('renders the recommendations form', async () => {
    const { findByRole, findByTestId } = renderForm('recommendations');

    const button = await findByRole('button', {
      name: 'Add another recommendation'
    });

    userEvent.click(button);

    // Title field
    const titleInput = await findByRole('textbox', { name: 'Title *' });
    userEvent.type(titleInput, mockRecommendation.title);
    expect(titleInput).toHaveValue(mockRecommendation.title);

    // Description field
    const descriptionInput = await findByRole('textbox', {
      name: 'Description *'
    });
    userEvent.type(descriptionInput, mockRecommendation.recommendation);
    expect(descriptionInput).toHaveValue(mockRecommendation.recommendation);

    // Add resource link
    const addLinkButton = await findByRole('button', {
      name: 'Add a resource link'
    });
    userEvent.click(addLinkButton);

    const addAnotherLinkButton = await findByRole('button', {
      name: 'Add another resource link'
    });
    // Button should be disabled while link input is blank
    expect(addAnotherLinkButton).toBeDisabled();

    let linkInput = await findByTestId('links.0.link');

    userEvent.type(linkInput, mockRecommendation.links[0]);
    expect(linkInput).toHaveValue(mockRecommendation.links[0]);

    userEvent.click(addAnotherLinkButton);

    linkInput = await findByTestId('links.1.link');

    const linkText = `https://www.${mockRecommendation.links[1]}`;
    userEvent.type(linkInput, linkText);
    expect(linkInput).toHaveValue(linkText);
  });

  it('renders the Next Steps form', async () => {
    const { findByRole, getByRole } = renderForm('next-steps');

    const nextStepsInput = await findByRole('textbox', {
      name: 'Next steps *'
    });

    expect(nextStepsInput).toHaveValue(adviceLetter.nextSteps);

    expect(
      getByRole('radio', {
        name: 'Yes, a follow-up is recommended'
      })
    ).toBeChecked();

    const followupPointInput = getByRole('textbox', { name: 'When?' });
    expect(followupPointInput).toHaveValue(adviceLetter.followupPoint);

    // Check that followup point input is hidden when followup radio field is false
    userEvent.click(getByRole('radio', { name: 'Not necessary' }));
    expect(followupPointInput).not.toBeInTheDocument();
  });
});
