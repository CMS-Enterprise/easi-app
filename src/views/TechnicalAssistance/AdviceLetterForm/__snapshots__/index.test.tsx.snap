// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TRB Advice Letter Form matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="trb-form-header"
  >
    <div
      class="bg-gray-5 padding-bottom-6"
    >
      <div
        class="grid-container"
        data-testid="gridContainer"
      >
        <div
          class="grid-row"
          data-testid="grid"
        >
          <div
            class="grid-col"
            data-testid="grid"
          >
            <nav
              aria-label="Breadcrumbs"
              class="usa-breadcrumb padding-bottom-0"
              style="background-color: transparent;"
            >
              <ol
                class="usa-breadcrumb__list"
              >
                <li
                  class="usa-breadcrumb__list-item"
                >
                  <a
                    class="usa-breadcrumb__link"
                    href="/trb"
                  >
                    <span>
                      Home
                    </span>
                  </a>
                </li>
                <li
                  class="usa-breadcrumb__list-item"
                >
                  <a
                    class="usa-breadcrumb__link"
                    href="/trb/d8ebedca-0031-4ccd-9690-37390726c50e/advice"
                  >
                    <span>
                      Request details
                    </span>
                  </a>
                </li>
                <li
                  aria-current="page"
                  class="usa-breadcrumb__list-item usa-current"
                >
                  <span>
                    Advice letter
                  </span>
                </li>
              </ol>
            </nav>
            <h1
              aria-live="polite"
              class="easi-h1 margin-bottom-0"
              tabindex="-1"
            >
              Advice letter
            </h1>
            <div
              class="font-body-lg line-height-body-5 text-light"
            >
              Document any outcomes, advice, findings, recommendations, or next steps that the TRB has for this request.
            </div>
            <div
              class="margin-top-2 line-height-body-5"
            >
              After submitting this form, the requester will recieve an automatic notification letting them know their advice letter is available.
            </div>
            <div
              class="margin-top-3"
            >
              <button
                class="usa-button usa-button--unstyled"
                data-testid="button"
                type="button"
              >
                <svg
                  class="usa-icon margin-right-05 margin-bottom-2px text-tbottom"
                  focusable="false"
                  height="1em"
                  role="img"
                  viewBox="0 0 24 24"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M0 0h24v24H0z"
                    fill="none"
                  />
                  <path
                    d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20v-2z"
                  />
                </svg>
                Save and return to request
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="grid-container"
      data-testid="gridContainer"
    >
      <div
        aria-label="progress"
        class="usa-step-indicator margin-top-4 margin-bottom-2"
        data-testid="step-indicator"
      >
        <ol
          class="usa-step-indicator__segments"
        >
          <li
            aria-current="true"
            class="usa-step-indicator__segment usa-step-indicator__segment--current usa-step-indicator__segment--disabled"
          >
            <span
              class="usa-step-indicator__segment-label"
            >
              <span
                class="name"
              >
                What we heard
              </span>
              <span
                class="long"
              >
                What we heard
              </span>
               
            </span>
          </li>
          <li
            class="usa-step-indicator__segment usa-step-indicator__segment--clickable"
          >
            <span
              class="usa-step-indicator__segment-label"
            >
              <span
                class="name"
              >
                What we recommend
              </span>
              <span
                class="long"
              >
                What we recommend
              </span>
               
              <span
                class="usa-sr-only"
              >
                not completed
              </span>
            </span>
          </li>
          <li
            class="usa-step-indicator__segment usa-step-indicator__segment--clickable"
          >
            <span
              class="usa-step-indicator__segment-label"
            >
              <span
                class="name"
              >
                Next steps
              </span>
              <span
                class="long"
              >
                Next steps
              </span>
               
              <span
                class="usa-sr-only"
              >
                not completed
              </span>
            </span>
          </li>
          <li
            class="usa-step-indicator__segment usa-step-indicator__segment--clickable"
          >
            <span
              class="usa-step-indicator__segment-label"
            >
              <span
                class="name"
              >
                Internal review
              </span>
              <span
                class="long"
              >
                Check your work and request an internal review
              </span>
               
              <span
                class="usa-sr-only"
              >
                not completed
              </span>
            </span>
          </li>
          <li
            class="usa-step-indicator__segment usa-step-indicator__segment--disabled"
          >
            <span
              class="usa-step-indicator__segment-label"
            >
              <span
                class="name"
              >
                Check and send
              </span>
              <span
                class="long"
              >
                Check the content of your advice letter and send it to the requester
              </span>
               
              <span
                class="usa-sr-only"
              >
                not completed
              </span>
            </span>
          </li>
        </ol>
        <div
          class="usa-step-indicator__header"
        >
          <h4
            class="usa-step-indicator__heading"
          >
            <span
              class="usa-step-indicator__heading-counter"
            >
              <span
                class="usa-sr-only"
              >
                Step
              </span>
              <span
                class="usa-step-indicator__current-step"
              >
                1
              </span>
               
              <span
                class="usa-step-indicator__total-steps"
              >
                of 5
              </span>
               
            </span>
            <span
              class="usa-step-indicator__heading-text"
            >
              <span
                class="name"
              >
                What we heard
              </span>
              <span
                class="long"
              >
                What we heard
              </span>
            </span>
          </h4>
        </div>
      </div>
      <div
        class="line-height-body-5 text-light"
      >
        Provide a brief but detailed summary of the purpose of the session, what was discussed or presented, and any other meeting details of note.
      </div>
    </div>
  </div>
  <div
    class="grid-container"
    data-testid="gridContainer"
  >
    <div
      class=""
      data-testid="grid"
    >
      <form
        class="usa-form maxw-tablet"
        data-testid="form"
        id="trbAdviceSummary"
      >
        <div
          class="easi-help-text margin-top-1 margin-bottom-1 text-base"
        >
          Fields marked with an asterisk ( 
          <span
            class="text-red"
          >
            *
          </span>
           ) are required.
        </div>
        <div
          class="usa-form-group"
          data-testid="formGroup"
        >
          <label
            class="usa-label text-normal"
            for="meetingSummary"
          >
            Meeting summary 
            <span
              class="text-red"
            >
               *
            </span>
          </label>
          <textarea
            class="usa-textarea"
            id="meetingSummary"
            name="meetingSummary"
          >
            Meeting summary text
          </textarea>
          <span
            aria-atomic="false"
            aria-live="polite"
            class="easi-character-counter margin-top-1"
            id="meetingSummary-counter"
          >
            1980 characters remaining
          </span>
        </div>
        <div
          class="margin-top-4"
        >
          <ul
            class="usa-button-group"
          >
            <li
              class="usa-button-group__item"
            >
              <div
                class="display-flex flex-align-center"
              >
                <button
                  class="usa-button usa-button--outline margin-top-0 margin-bottom-1 mobile-lg:margin-bottom-0"
                  data-testid="button"
                  type="button"
                >
                  Cancel
                </button>
              </div>
            </li>
            <li
              class="usa-button-group__item"
            >
              <div
                class="display-flex flex-align-center"
              >
                <button
                  class="usa-button margin-top-0"
                  data-testid="button"
                  type="submit"
                >
                  Next
                </button>
              </div>
            </li>
          </ul>
          <button
            class="usa-button usa-button--unstyled margin-top-2 display-flex flex-align-center"
            data-testid="button"
            type="button"
          >
            <svg
              class="usa-icon margin-right-05"
              focusable="false"
              height="1em"
              role="img"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
              <path
                d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20v-2z"
              />
            </svg>
            Save and return to request
          </button>
        </div>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
